plugins/multilanguage.ts

```diff
@@ -176,7 +183,56 @@ export default function multilanguage(userOptions: Options) {
           document.head.appendChild(meta);
           document.head.appendChild(document.createTextNode("\n"));
         }
+
+        if (page.data.alternateX) {
+          addHreflangX(page.data.alternateX, site, document);
+        }
       }
     });
   };
 }
+
+function getAlternateXData(
+  currentPage: Page<Data>,
+  validList: string[],
+  filteredPages: Page<Data>[],
+): Data | undefined {
+  const langX = currentPage.data.langX;
+
+  if (!langX) return void 0;
+
+  // Use case when langX is URL string
+  if (langX.startsWith("/")) {
+    const regUrl = new RegExp(`^\/${trim(langX, "\/")}\/?`, "i");
+    return filteredPages.filter(({ data }) => {
+      data.basename;
+      const [isMatch] = data.url.match(regUrl) ?? [];
+      return !!isMatch;
+    })
+      ?.[0]?.data;
+  }
+
+  // Ensure the fallback language is declared in the config file.
+  if (!validList.includes(langX)) return void 0;
+
+  return filteredPages.filter(({ data }) => {
+    const notSameContentRoot = data.id !== currentPage.data.id;
+    if (notSameContentRoot) return false;
+    const isXLang = data.lang === langX;
+    return isXLang;
+  })?.[0].data;
```

Comment:
It's better to get this once the alternates object is defined. So you can do simply:

```js
const alternateX = page.data.alternates.find((data) => data.lang === langX);
```

Reply:
The reason I did this is group the meaning of code into 1 block rather than let it all over the places. But your solution seems to have better performance. I'll think more about this.